crates/typst-pdf/src/page.rs

```diff
@@ -711,6 +724,129 @@ fn write_group(ctx: &mut PageContext, pos: Point, group: &GroupItem) {
 
 /// Encode a text run into the content stream.
 fn write_text(ctx: &mut PageContext, pos: Point, text: &TextItem) {
+    // If the text run contains either only emojis or normal text
+    // we can render it directly
+    let sbix = match text.font.ttf().tables().sbix {
+        None => {
+            write_normal_text(ctx, pos, text);
+            return;
+        }
+        Some(sbix) => sbix,
+    };
+
+    let strike = sbix.best_strike(text.size.to_f32() as u16).unwrap();
+    let is_emoji = |g: &Glyph| strike.get(GlyphId(g.id)).is_some();
+    let emoji_count = text.glyphs.iter().filter(|g| is_emoji(*g)).count();
+
+    if emoji_count == text.glyphs.len() {
+        write_emojis(ctx, pos, text);
+    } else if emoji_count == 0 {
+        write_text(ctx, pos, text)
+    } else {
+        // Otherwise we need to split it in smaller text runs
```

Comment:
Maybe this should be done ahead, so that at this point we have the guarantee that a text run is either all emojis or all non-emojis? I don't know the different passes of the compiler well enough to be able to tell where it could go or if it should stay here.

Reply:
I think doing it here in the PDF export is reasonable since it's an idiosyncracy of PDF that it needs to be split at all.